src/lib.rs

```diff
@@ -1258,30 +1121,109 @@ pub enum Descriptor {
 #[repr(C)]
 pub struct File {
     fd: wasi_filesystem::Descriptor,
-    position: u64,
+    position: Cell<u64>,
 }
 
 const PAGE_SIZE: usize = 65536;
 
-impl Descriptor {
-    fn get(fd: Fd) -> &'static mut Descriptor {
+/// The maximum path length. WASI doesn't explicitly guarantee this, but all
+/// popular OS's have a `PATH_MAX` of at most 4096, so that's enough for this
+/// polyfill.
+const PATH_MAX: usize = 4096;
+
+struct State {
+    buffer_ptr: Cell<*mut u8>,
+    buffer_len: Cell<usize>,
+    ndescriptors: usize,
+    descriptors: MaybeUninit<[Descriptor; 128]>,
+    path_buf: MaybeUninit<[u8; PATH_MAX]>,
+}
+
+extern "C" {
+    fn get_global_ptr() -> usize;
+    fn set_global_ptr(a: usize);
+}
+
+impl State {
+    fn with(f: impl FnOnce(&State) -> Result<(), Errno>) -> Errno {
+        let ptr = State::ptr();
+        let ptr = ptr.try_borrow().unwrap_or_else(|_| unreachable());
+        let ret = f(&*ptr);
+        match ret {
+            Ok(()) => ERRNO_SUCCESS,
+            Err(err) => err,
+        }
+    }
+
+    fn with_mut(f: impl FnOnce(&mut State) -> Result<(), Errno>) -> Errno {
+        let ptr = State::ptr();
+        let mut ptr = ptr.try_borrow_mut().unwrap_or_else(|_| unreachable());
+        let ret = f(&mut *ptr);
+        match ret {
+            Ok(()) => ERRNO_SUCCESS,
+            Err(err) => err,
+        }
+    }
+
+    fn ptr() -> &'static RefCell<State> {
+        assert!(size_of::<State>() <= PAGE_SIZE);
         unsafe {
-            let mut fds = replace_fds(null_mut());
-            if fds.is_null() {
-                let grew = core::arch::wasm32::memory_grow(0, 1);
-                if grew == usize::MAX {
-                    unreachable();
+            match get_global_ptr() {
+                0 => {
+                    let grew = core::arch::wasm32::memory_grow(0, 1);
+                    if grew == usize::MAX {
+                        unreachable();
+                    }
+                    let ret = (grew * PAGE_SIZE) as *mut RefCell<State>;
+                    set_global_ptr(ret as usize);
+                    ret.write(RefCell::new(State {
+                        buffer_ptr: Cell::new(null_mut()),
+                        buffer_len: Cell::new(0),
+                        ndescriptors: 0,
+                        descriptors: MaybeUninit::uninit(),
+                        path_buf: MaybeUninit::uninit(),
+                    }));
+                    &*ret
                 }
-                fds = (grew * PAGE_SIZE) as *mut u8;
+                1 => unreachable(),
```

Comment:
Can you add a comment here about what  `get_global_ptr` returning 1 would mean?

Reply:
Oops thank for catching this, it's a vestige of a prior design which I didn't settle on so just needs removal now.